eloquentjs/4_2_reverse_array.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * Performs non-mutating array reversal operation
+ * @param {*} input 
+ * @returns 
+ */
+function reverseArray(input) {
+    if (!Array.isArray(input)) {
+        throw "input must be array";
+    }
+
+    output = [];
+    for (let i = 0; i < input.length; i++) {
+        output.unshift(input[i]);
+    }
+    return output;
+}
+
+/**
+ * Performs mutating array reversal operation
+ * @param {*} input 
+ * @returns 
+ */
+function reverseArrayInPlace(input) {
+    if (!Array.isArray(input)) {
+        throw "input must be array";
+    }
+
+    for (let i = 0; i < input.length - 1; i++) {
+        input.splice(i,0,input.pop())
```

Comment:
Using splice with a deleteCount of 0 inserts the value at index i which shifts every element of the array after i forward, making this an O(n^2) solution instead of an O(n) solution. Is there a solution that's O(n)? Do you know how to use a swap variable?